File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/IFInventoryListener.java
package me.devnatan.inventoryframework;

import me.devnatan.inventoryframework.context.IFCloseContext;
import me.devnatan.inventoryframework.context.IFContext;
import me.devnatan.inventoryframework.context.IFRenderContext;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.inventory.InventoryDragEvent;
import org.bukkit.event.player.PlayerDropItemEvent;
import org.bukkit.event.player.PlayerPickupItemEvent;
import org.bukkit.event.server.PluginDisableEvent;

@SuppressWarnings("unused")
final class IFInventoryListener implements Listener {

    private final ViewFrame viewFrame;

    public IFInventoryListener(ViewFrame viewFrame) {
        this.viewFrame = viewFrame;
    }

    @EventHandler
    public void onPluginDisable(final PluginDisableEvent event) {
        if (!event.getPlugin().getName().equals(viewFrame.getOwner().getName())) return;

        viewFrame.unregister();
    }

    @EventHandler(ignoreCancelled = true)
    public void onInventoryClick(final InventoryClickEvent event) {
        if (!(event.getWhoClicked() instanceof Player)) return;

        final Player player = (Player) event.getWhoClicked();
        final Viewer viewer = viewFrame.getViewer(player);
        if (viewer == null) return;

        viewer.getActiveContext().simulateClick(event.getRawSlot(), viewer, event, false);
    }

    @EventHandler(ignoreCancelled = true)
    public void onInventoryClose(final InventoryCloseEvent event) {
        if (!(event.getPlayer() instanceof Player)) return;

        final Player player = (Player) event.getPlayer();
        final Viewer viewer = viewFrame.getViewer(player);
        if (viewer == null) return;

        final IFRenderContext context = viewer.getActiveContext();
        final RootView root = context.getRoot();
        final IFCloseContext closeContext = root.getElementFactory().createCloseContext(viewer, context);

        closeContext.simulateCloseForPlayer();
    }

    @SuppressWarnings("deprecation")
    @EventHandler(ignoreCancelled = true)
    public void onItemPickup(final PlayerPickupItemEvent event) {
        final Viewer viewer = viewFrame.getViewer(event.getPlayer());
        if (viewer == null) return;

        final IFContext context = viewer.getActiveContext();
        if (!context.getConfig().isOptionSet(ViewConfig.CANCEL_ON_PICKUP)) return;

        event.setCancelled(context.getConfig().getOptionValue(ViewConfig.CANCEL_ON_PICKUP));
    }

    @EventHandler(ignoreCancelled = true)
    public void onItemDrop(final PlayerDropItemEvent event) {
        final Viewer viewer = viewFrame.getViewer(event.getPlayer());
        if (viewer == null) return;

        final IFContext context = viewer.getActiveContext();
        if (!context.getConfig().isOptionSet(ViewConfig.CANCEL_ON_DROP)) return;

        event.setCancelled(context.getConfig().getOptionValue(ViewConfig.CANCEL_ON_DROP));
    }

    @EventHandler(ignoreCancelled = true)
    public void onInventoryDrag(final InventoryDragEvent event) {
        if (!(event.getWhoClicked() instanceof Player)) return;

        final Viewer viewer = viewFrame.getViewer((Player) event.getWhoClicked());
        if (viewer == null) return;

        final IFContext context = viewer.getActiveContext();
        if (!context.getConfig().isOptionSet(ViewConfig.CANCEL_ON_DRAG)) return;

        final boolean configValue = context.getConfig().getOptionValue(ViewConfig.CANCEL_ON_DRAG);
        final int size = event.getInventory().getSize();
        for (final int rawSlot : event.getRawSlots()) {
            if (!(rawSlot < size)) continue;

            event.setCancelled(configValue);
            break;
        }
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/IFInventoryListener.kt
package me.devnatan.inventoryframework

import me.devnatan.inventoryframework.context.IFCloseContext
import me.devnatan.inventoryframework.context.IFContext
import me.devnatan.inventoryframework.context.IFRenderContext
import me.devnatan.inventoryframework.context.IFSlotClickContext
import me.devnatan.inventoryframework.pipeline.StandardPipelinePhases
import net.minestom.server.entity.Player
import net.minestom.server.event.EventFilter
import net.minestom.server.event.EventNode
import net.minestom.server.event.inventory.InventoryCloseEvent
import net.minestom.server.event.inventory.InventoryPreClickEvent
import net.minestom.server.event.item.PickupItemEvent
import net.minestom.server.event.trait.EntityEvent
import net.minestom.server.inventory.PlayerInventory
import net.minestom.server.inventory.click.ClickType
import kotlin.jvm.optionals.getOrNull

internal class IFInventoryListener(
    private val viewFrame: ViewFrame,
    handler: EventNode<in EntityEvent>,
) {
    init {
        val node =
            EventNode
                .type("IF", EventFilter.ENTITY) { _, e ->
                    e is Player && viewFrame.getViewer(e) != null
                }.setPriority(10)
                .addListener(PickupItemEvent::class.java, this::onItemPickup)
                .addListener(InventoryPreClickEvent::class.java, this::onInventoryClick)
                .addListener(InventoryCloseEvent::class.java, this::onInventoryClose)
        handler.addChild(node)
    }

    fun onInventoryClick(event: InventoryPreClickEvent) {
        if (event.isCancelled) return

        val player = event.player
        val viewer = viewFrame.getViewer(player) ?: return

        if (event.clickType == ClickType.DROP) {
            val context: IFContext = viewer.activeContext
            if (!context.config.isOptionSet(ViewConfig.CANCEL_ON_DROP)) return

            event.isCancelled = context.config.getOptionValue(ViewConfig.CANCEL_ON_DROP)
            return
        }
        if (
            event.clickType == ClickType.LEFT_DRAGGING ||
            event.clickType == ClickType.RIGHT_DRAGGING
        ) {
            val context: IFContext = viewer.activeContext
            if (!context.config.isOptionSet(ViewConfig.CANCEL_ON_DRAG)) return

            event.isCancelled = context.config.getOptionValue(ViewConfig.CANCEL_ON_DRAG)
            return
        }

        val context: IFRenderContext = viewer.activeContext
        val clickedComponent =
            context
                .getComponentsAt(event.slot)
                .stream()
                .filter { it.isVisible }
                .findFirst()
                .getOrNull()
        val clickedContainer =
            if (event.inventory is PlayerInventory) {
                viewer.selfContainer
            } else {
                context.getContainer()
            }

        val root: RootView = context.getRoot()
        val clickContext: IFSlotClickContext =
            root.elementFactory.createSlotClickContext(
                event.slot,
                viewer,
                clickedContainer,
                clickedComponent,
                event,
                false,
            )

        root.pipeline.execute(StandardPipelinePhases.CLICK, clickContext)
    }

    fun onInventoryClose(event: InventoryCloseEvent) {
        val player: Player = event.player
        val viewer = viewFrame.getViewer(player) ?: return

        val context: IFRenderContext = viewer.activeContext
        val root: RootView = context.getRoot()
        val closeContext: IFCloseContext = root.elementFactory.createCloseContext(viewer, context)

        root.pipeline.execute(StandardPipelinePhases.CLOSE, closeContext)
    }

    fun onItemPickup(event: PickupItemEvent) {
        val viewer = viewFrame.getViewer(event.entity as Player) ?: return

        val context: IFContext = viewer.activeContext
        if (!context.getConfig().isOptionSet(ViewConfig.CANCEL_ON_PICKUP)) return

        event.isCancelled = context.getConfig().getOptionValue(ViewConfig.CANCEL_ON_PICKUP)
    }
}
